experimental/devices/cap1188/cap1188.go

```diff
@@ -0,0 +1,554 @@
+	if err != nil {
+		return inputStatuses, wrap(fmt.Errorf("failed to read the input values - %s", err))
+	}
+
+	// read deltas (in two's complement, capped at -128 to 127)
+	deltasB := [nbrOfLEDs]byte{}
+	if err = d.regWrapper.ReadStruct(0x10, &deltasB); err != nil {
+		return inputStatuses, wrap(fmt.Errorf("failed to read the delta values - %s", err))
+	}
+	deltas := [nbrOfLEDs]int{}
+	for i, b := range deltasB {
+		deltas[i] = twoComplementsToInt(b)
+	}
+
+	// read threshold
+	thresholds := [nbrOfLEDs]byte{}
+	if err = d.regWrapper.ReadStruct(0x30, &thresholds); err != nil {
+		return inputStatuses, wrap(fmt.Errorf("failed to read the threshold values - %s", err))
+	}
+
+	// convert the data into a sensor state
+	var touched bool
+	for i := uint8(0); i < nbrOfLEDs; i++ {
+		touched = isBitSet(status, 7-i)
+
+		// TODO: check if the event is passed the threshold:
+		// deltas[i] > int(thresholds[i])
+
+		if touched {
+			if inputStatuses[i] == PressedStatus {
+				if d.opts.RetriggerOnHold {
+					inputStatuses[i] = HeldStatus
+				}
+				continue
+			}
+			inputStatuses[i] = PressedStatus
+		} else {
+			inputStatuses[i] = OffStatus
+		}
+	}
+
+	return inputStatuses, nil
+}
+
+// LinkLeds link the behavior of the LEDs to the touch sensors.
+// Doing so, disabled the option for the host to set specific LEDs on/off.
+func (d *Dev) LinkLeds() error {
+	if err := d.regWrapper.WriteUint8(reg_LEDLinking, 0xff); err != nil {
+		return wrap(fmt.Errorf("failed to link LEDs - %s", err))
+	}
+	d.opts.LinkedLEDs = true
+	return nil
+}
+
+// UnlinkLeds disassociate the LEDs from the input sensors allowing the host to
+// control the LEDs.
+func (d *Dev) UnlinkLeds() error {
+	if err := d.regWrapper.WriteUint8(reg_LEDLinking, 0x00); err != nil {
+		return wrap(fmt.Errorf("failed to unlink LEDs - %s", err))
+	}
+	d.opts.LinkedLEDs = false
+	return nil
+}
+
+// AllLedsOn turns all the LEDs on.
+// This is quite more efficient than looping through each led and turn them on.
+func (d *Dev) AllLedsOn() error {
+	if d.opts.LinkedLEDs {
+		return wrap(fmt.Errorf("can't manually set LEDs when they are linked to sensors"))
+	}
+	return d.regWrapper.WriteUint8(reg_LEDOutputControl, 0xff)
+}
+
+// AllLedsOff turns all the LEDs off.
+// This is quite more efficient than looping through each led and turn them off.
+func (d *Dev) AllLedsOff() error {
+	if d.opts.LinkedLEDs {
+		return wrap(fmt.Errorf("can't manually set LEDs when they are linked to sensors"))
+	}
+	return d.regWrapper.WriteUint8(reg_LEDOutputControl, 0x00)
+}
+
+// SetLed sets the state of a LED as on or off
+// Only works if the LEDs are not linked to the sensors
+func (d *Dev) SetLed(idx int, state bool) error {
+	if d.opts.LinkedLEDs {
+		return wrap(fmt.Errorf("can't manually set LEDs when they are linked to sensors"))
+	}
+	if idx > 7 || idx < 0 {
+		return wrap(fmt.Errorf("invalid led idx %d", idx))
+	}
+	if Debug {
+		fmt.Printf("Set LED state %d - %t\n", idx, state)
```

Comment:
Use log.Printf(). And I think it should be part of the Opts.

But that raises an interesting point about potentially interesting logging lines for devices drivers.

Also, somehow I think using an Opts struct is the best long term way to go for most device driver. Not sure why it took me a year to decide.

Reply:
Makes total sense